File: swift/bestTimeToBuyAndSellStock.swift
func maxProfit(prices: [Int]) -> Int {
    var profit = 0
    var hasStock = false
    var b = 0
    for i in 0..<prices.count {
        let currentP = prices[i]
        if i + 1 >= prices.count{
            if hasStock && currentP - b > 0 {
                profit += currentP - b
            }
            break
        }
        let nextP = prices[i+1]
        if hasStock {
            
            let currentProfit = currentP - b
            let futureProfit = nextP - b
            
            if futureProfit < currentProfit {
                // sell stock
                profit += currentProfit
                hasStock = false
            }
        } else {
            if nextP > currentP {
                // buy stock
                b = currentP
                hasStock = true
            }
        }
    }
    return profit
}

File: ruby/best_time_to_buy_and_sell_stock.rb
def max_profit(prices)
   profit = 0
   has_stock = false
   b = 0
   prices.each_with_index do |current_price, index|
       if index + 1 >= prices.count
           if has_stock && current_price - b > 0
               profit += current_price - b
           end
           break
       end
       next_price = prices[index+1]
       if has_stock 
           current_profit = current_price - b
           future_profit = next_price - b
           if future_profit < current_profit
               profit += current_profit
               has_stock = false
           end
       else
           if next_price > current_price
               b = current_price
               has_stock = true
           end
       end
   end
   return profit
end
